File: apps/api-gql/internal/delivery/gql/mappers/events.go
package mappers

import (
	"github.com/twirapp/twir/apps/api-gql/internal/delivery/gql/gqlmodel"
	"github.com/twirapp/twir/apps/api-gql/internal/entity"
)

func MapEventToGQL(event entity.Event) gqlmodel.Event {
	operations := make([]gqlmodel.EventOperation, 0, len(event.Operations))
	for _, op := range event.Operations {
		filters := make([]gqlmodel.EventOperationFilter, 0, len(op.Filters))
		for _, f := range op.Filters {
			filters = append(
				filters, gqlmodel.EventOperationFilter{
					ID:    f.ID,
					Type:  f.Type,
					Left:  f.Left,
					Right: f.Right,
				},
			)
		}

		operations = append(
			operations, gqlmodel.EventOperation{
				ID:             op.ID,
				Type:           op.Type,
				Input:          op.Input,
				Delay:          op.Delay,
				Repeat:         op.Repeat,
				UseAnnounce:    op.UseAnnounce,
				TimeoutTime:    op.TimeoutTime,
				TimeoutMessage: op.TimeoutMessage,
				Target:         op.Target,
				Enabled:        op.Enabled,
				Filters:        filters,
			},
		)
	}

	return gqlmodel.Event{
		ID:          event.ID,
		ChannelID:   event.ChannelID,
		Type:        event.Type,
		RewardID:    event.RewardID,
		CommandID:   event.CommandID,
		KeywordID:   event.KeywordID,
		Description: event.Description,
		Enabled:     event.Enabled,
		OnlineOnly:  event.OnlineOnly,
		Operations:  operations,
	}
}

File: frontend/dashboard/src/api/events.ts
import { useQuery } from '@urql/vue'
import { createGlobalState } from '@vueuse/core'

import type { Event as GqlEvent, EventOperation as GqlEventOperation } from '@/gql/graphql'

import { useMutation } from '@/composables/use-mutation'
import { graphql } from '@/gql'

export type Event = GqlEvent
export type EventOperation = GqlEventOperation
export { EventType } from '@/gql/graphql'

const invalidationKey = 'Events'

export const useEventsApi = createGlobalState(() => {
	// Query all events
	const useQueryEvents = () => useQuery({
		query: graphql(`
      query GetEvents {
        events {
          id
          channelId
          type
          rewardId
          commandId
          keywordId
          description
          enabled
          onlineOnly
          operations {
            id
            type
            input
            delay
            repeat
            useAnnounce
            timeoutTime
            timeoutMessage
            target
            enabled
            filters {
              id
              type
              left
              right
            }
          }
        }
      }
    `),
		context: {
			additionalTypenames: [invalidationKey],
		},
	})

	// Query a single event by ID
	const useQueryEventById = (id: string) => useQuery({
		query: graphql(`
      query GetEventById($id: String!) {
        eventById(id: $id) {
          id
          channelId
          type
          rewardId
          commandId
          keywordId
          description
          enabled
          onlineOnly
          operations {
            id
            type
            input
            delay
            repeat
            useAnnounce
            timeoutTime
            timeoutMessage
            target
            enabled
            filters {
              id
              type
              left
              right
            }
          }
        }
      }
    `),
		variables: { id },
		context: {
			additionalTypenames: [invalidationKey],
		},
		pause: true,
	})

	// Create a new event
	const useMutationCreateEvent = () => useMutation(
		graphql(`
      mutation CreateEvent($input: EventCreateInput!) {
        eventCreate(input: $input) {
          id
        }
      }
    `),
		[invalidationKey],
	)

	// Update an existing event
	const useMutationUpdateEvent = () => useMutation(
		graphql(`
      mutation UpdateEvent($id: String!, $input: EventUpdateInput!) {
        eventUpdate(id: $id, input: $input) {
          id
        }
      }
    `),
		[invalidationKey],
	)

	// Delete an event
	const useMutationDeleteEvent = () => useMutation(
		graphql(`
      mutation DeleteEvent($id: String!) {
        eventDelete(id: $id)
      }
    `),
		[invalidationKey],
	)

	// Enable or disable an event
	const useMutationEnableOrDisableEvent = () => useMutation(
		graphql(`
      mutation EnableOrDisableEvent($id: String!, $enabled: Boolean!) {
        eventEnableOrDisable(id: $id, enabled: $enabled) {
          id
          enabled
        }
      }
    `),
		[invalidationKey],
	)

	return {
		useQueryEvents,
		useQueryEventById,
		useMutationCreateEvent,
		useMutationUpdateEvent,
		useMutationDeleteEvent,
		useMutationEnableOrDisableEvent,
	}
})
